Add docs about theming and customization

tomalec opened the pull request:
of Uniform Components and Solution's design
https://github.com/Starcounter/uniform.css/issues/80

tomalec committed:
Add docs about theming and customization

of Uniform Components and Solution's design
https://github.com/Starcounter/uniform.css/issues/80

tomalec committed:
Fix some grammar,

in the theming docs,
https://github.com/Starcounter/uniform.css/issues/80

ErlendLandro commented on docs/topic-guides/blendable-web-apps/starcounter-design-system.md:
```diff
@@ -2,22 +2,22 @@
-2. Uniform.css - UI patterns in a separated presentation layer
-3. Component library based on custom elements
-4. Themes based on CSS custom properties
+2. Uniform - a library of UI patterns and custom elements to be used in a separated presentation layer
```
Passive form, use "to use" over "to be used"

ErlendLandro commented on docs/topic-guides/blendable-web-apps/starcounter-design-system.md:
```diff
@@ -2,22 +2,22 @@
-3. Component library based on custom elements
-4. Themes based on CSS custom properties
+2. Uniform - a library of UI patterns and custom elements to be used in a separated presentation layer
+3. Customization through Themes and CSS custom properties
```
Is "Themes" intentionally capitalized?

tomalec replied:
I'm still not sure if we should use Themes capitalized, they are slightly specific things, so it's not a regular CSS file as any other, but something that meets our precise needs.

ErlendLandro replied:
It should only be capitalized if it's a name. If you mean something more specific when you write "theme", you should define it in the document.

ErlendLandro commented on docs/topic-guides/blendable-web-apps/starcounter-design-system.md:
```diff
@@ -2,22 +2,22 @@
-4. Themes based on CSS custom properties
+2. Uniform - a library of UI patterns and custom elements to be used in a separated presentation layer
+3. Customization through Themes and CSS custom properties
+4. Free, hackable world of HTML, CSS & JS
```
I don't understand how this fits in. Also, prefer "and" over "&"

tomalec replied:
Do you mean, you don't get how "Free,hackable.." fits to the list of layers? I wanted to express that with Starcounter, we are giving tools and conventions to support you, but we are not limiting developer's freedom, and at the and he/she is not bound to Starcounter specific language, API or elements, but can use any Web Platform technique to deliver view's presentation.

ErlendLandro replied:
Yes, exactly. I thought it was confusing when I read it. Again, you'd have to check with the app developers what they think

ErlendLandro commented on docs/topic-guides/blendable-web-apps/starcounter-design-system.md:
```diff
@@ -31,31 +31,46 @@ Underwear.css is configurable using CSS custom properties.
+More advanced components, like date pickers, are [implemented as custom elements](https://github.com/Starcounter/uniform.css/tree/master/components). They are meant to be used in shadow roots as well, to wrap distributed native HTML elements. That allows to serve real progressive enhancement, and what is crucial for a system of blended apps: keep app content's markup semantic and unopinionated, g ...
 
-Uniform.css is configurable using CSS custom properties.
+Uniform.css or Uniform components must be explicitly imported in any view shadow root that uses it.
```
Why do we say "Uniform.css" here and just "Uniform" in other places? What's the difference?

tomalec replied:
Currently Uniform consists of Uniform.css and Uniform components, repo and package name is still `uniform.css`. (@warpech  is this the time to finally rename it to the Unifrom?)

warpech replied:
Yes, let's rename to just Uniform.

ErlendLandro commented on docs/topic-guides/blendable-web-apps/starcounter-design-system.md:
```diff
@@ -31,31 +31,46 @@ Underwear.css is configurable using CSS custom properties.
+Uniform.css or Uniform components must be explicitly imported in any view shadow root that uses it.
 
-Head to the [demo page of Uniform.css](https://starcounter.github.io/uniform.css/) for a preview.
+Uniform is configurable using CSS custom properties and themes, for more, read [next section](#customization-through-themes-and-css-custom-properties).
```
"read next section" -> "read the next section"

ErlendLandro commented on docs/topic-guides/blendable-web-apps/starcounter-design-system.md:
```diff
@@ -31,31 +31,46 @@ Underwear.css is configurable using CSS custom properties.
+One of the tools in the [Blending](https://github.com/Starcounter/Blending) apps suite is a CSS custom properties manager, that stores color schemes in the database. The CSS custom properties can be stored in collections and provided conditionally in specific layouts.
+
+For more sophisticated theming we would love to use native [CSS shadow parts and themes](https://meowni.ca/posts/part-theme-explainer/), unfortunately, it's not supported yet.
+Luckily, we are able to mimic its behavior to the satisfactory degree with Themes tailored to our elements, based on [Vaadin themes](https://vaadin.com/themes).
```
Again, why is "Themes" capitalized? 

Does "to the satisfactory degree" add anything here? What would happen if we remove it?

tomalec replied:
To be honest we are far from mimicking its behavior as a spec. But we are mimicking only a part we need to the extent of our use-case.

ErlendLandro replied:
Would that be possible to explain that then? "to the satisfactory degree" doesn't tell me anything

ErlendLandro commented on docs/topic-guides/blendable-web-apps/starcounter-design-system.md:
```diff
@@ -31,31 +31,46 @@ Underwear.css is configurable using CSS custom properties.
+
+For more sophisticated theming we would love to use native [CSS shadow parts and themes](https://meowni.ca/posts/part-theme-explainer/), unfortunately, it's not supported yet.
+Luckily, we are able to mimic its behavior to the satisfactory degree with Themes tailored to our elements, based on [Vaadin themes](https://vaadin.com/themes).
+The main difference is that those Themes works on the global level, but given that you can still customize them per each view level - it's the typical case anyway.
```
I don't understand this sentence. the "but given" doesn't seem to have an ending. 

Also, "customize per view level" or "customize each view level" not "customize them per each view level"

tomalec replied:
W3C `::theme` could be "scoped" as any other CSS selector, you can do `element::theme(foo)` which will select and style(theme) all `foo` parts that are descendants of `element`.

Our themes work only globally, so when you want to theme `foo` part you can only do it for all `part`s in the entire document.

I believe it's not so problematic. When you want to customize it on per view level, you can use regular compositions (however then you are bound to this level only, no recursion or cascading). That's why I assume, the typical use case for themes is to actually style **all** elements

ErlendLandro replied:
Okay, I think I might understand what you're saying. I think the "but" should be removed then: 
> The main difference is that those Themes works on the global level, given that you can still customize them per view level

ErlendLandro commented on docs/topic-guides/blendable-web-apps/starcounter-design-system.md:
```diff
@@ -31,31 +31,46 @@ Underwear.css is configurable using CSS custom properties.
+
+We also have a GUI tool in the [Blending](https://github.com/Starcounter/Blending) apps suite to manage themes for specific layouts.
+
+To read more on customizing the design of the blended solution, please check [dedicated page](customizing-solutions-design.md).
```
"the dedicated page"

ErlendLandro commented on docs/topic-guides/blendable-web-apps/customizing-solutions-design.md:
```diff
@@ -0,0 +1,92 @@
+
+Building your app or solution design the way we suggest does not only help you create a consistent look and feel fast and easily. It also gives you means to customize it later, by applying global themes, switching color & configuration schemes, or replacing the definition of every component at once.
+
+This page covers ways of customizing your **entire** app's  (solution's) design at once. You can always customize/edit every single view individually, by changing its composition, see [View Composing](view-composing.md)
```
"app or solution's design at once", using parentheses this way makes it hard to read

"customize" or "edit", not "customize/edit", using one word makes it more specific and easier to read

Words like "always" and "single" doesn't add information and makes the text wordy

ErlendLandro commented on docs/topic-guides/blendable-web-apps/customizing-solutions-design.md:
```diff
@@ -0,0 +1,92 @@
+
+## Themes
+
+Theming is the very high-level way to change the look and feel of many UI elements and patterns with a single step, by just attaching already created theme to given layout.
```
"very" is an unnecessary word here 

Do you think it would work to shorten the end of this sentence:

> change the look and feel of UI elements and patterns by attaching themes to layouts.

tomalec replied:
I would like to emphasize that i'ts a single simple step, you take theme from somewhere and just attach it to the existing layout.

I agree my sentence is hard to read:| and like to readability of yours, but from it, I got a feeling that "attaching themes to layouts" takes many themes and layouts and attaching is a process that is a bit longer.

ErlendLandro commented on docs/topic-guides/blendable-web-apps/customizing-solutions-design.md:
```diff
@@ -0,0 +1,92 @@
+
+This requires an element to actively look up for a theme when it's defined. That's how Vaadin custom elements work, and that's how we will keep Uniform Components working. Therefore, as long as you stick to [Starcounter Design System](starcounter-design-system.md), you don't need to worry about those implementation details.
+
+#### But why good old Bootstrap-like themes do not work?
```
Why are we suddenly on an H4 level here? The parent is H2.

Can you take a look at the [Microsoft Style Guide for headlines](https://docs.microsoft.com/en-us/style-guide/scannable-content/headings) and see if you can improve these headings?

tomalec replied:
> Why are we suddenly on an H4 level here? The parent is H2.

It's a leftover of my initial idea of different structure.

tomalec replied:
BTW, do we hane "side note" kind of hint? I think this section is sligthly side tracking.

ErlendLandro commented on docs/topic-guides/blendable-web-apps/customizing-solutions-design.md:
```diff
@@ -0,0 +1,92 @@
+
+#### How to use it?
+
+If you are not into the code, you can start the [Blending](https://github.com/Starcounter/Blending) apps suite, and use a [GUI to manage the themes](https://github.com/Starcounter/Blending/blob/master/docs/gui.md). There you can pick a layout, then add URLs to the themes you would like to add.
```
What do you mean with "If you are not into the code"? 

"app suite" is implicitly plural, so there's no reason to write "apps suite"

tomalec replied:
I want to give separate hints for solution owners not interested in coding, and solution owners that want to have fine-tuned results and actually want to have full control with code.

ErlendLandro commented on docs/topic-guides/blendable-web-apps/customizing-solutions-design.md:
```diff
@@ -0,0 +1,92 @@
+
+## Collections of CSS Custom Properties
+
+Slightly lower-level way to customize existing design is by using CSS Variables (Custom Properties). Well-behaved design systems, themes, custom elements and even individual CSS rules usually use CSS Variables which could be used to customize it.
```
"Slightly" is an unnecessary word here, better to skip it:

> A lower-level way...

ErlendLandro commented on docs/topic-guides/blendable-web-apps/customizing-solutions-design.md:
```diff
@@ -0,0 +1,92 @@
+
+Slightly lower-level way to customize existing design is by using CSS Variables (Custom Properties). Well-behaved design systems, themes, custom elements and even individual CSS rules usually use CSS Variables which could be used to customize it.
+
+This allows to change color schemes, or configure other values like font sizing, margins, etc.
```
Why is this in its own paragraph?

tomalec replied:
In every section (Themes, Variables, Elements) I kept the structure where second paragraph shortly state what each method allows you to do, and when to use it.
While the first describes what it is and how it briefly works.

ErlendLandro commented on docs/topic-guides/blendable-web-apps/customizing-solutions-design.md:
```diff
@@ -0,0 +1,92 @@
+
+This allows to change color schemes, or configure other values like font sizing, margins, etc.
+
+Again [Blending](https://github.com/Starcounter/Blending) apps suite [delivers GUI](https://github.com/Starcounter/Blending/blob/master/docs/gui.md#css-variables-manager) to do that without touching any code. You can as well add it to you composition code manually.
```
"to do that without touching any code". I don't know what "that" refers to here

ErlendLandro commented on docs/topic-guides/blendable-web-apps/customizing-solutions-design.md:
```diff
@@ -0,0 +1,92 @@
+
+Again [Blending](https://github.com/Starcounter/Blending) apps suite [delivers GUI](https://github.com/Starcounter/Blending/blob/master/docs/gui.md#css-variables-manager) to do that without touching any code. You can as well add it to you composition code manually.
+
+{% hint style="warning" %} Attaching CSS Variables' collection through Blending app, allows you to set them on light DOM. Currently to the global scope - `body` (https://github.com/Starcounter/Blending/issues/232). While setting it in view's composition, attaches them in Shadow DOM to the scope you choose. {% endhint %}
```
This hint is hard to understand. Do you think you could clarify?

tomalec replied:
Actually I'd love to clarify and change the implementation https://github.com/Starcounter/Blending/issues/232 so it could be totally removed as both will apply to compositions only, or in case we need to apply collections to light for some reason, just:

> Attaching CSS Variables' collection through Blending app, allows you to set them on light DOM for selected view (will cascade to all subviews). While setting it manually in view's composition, allows you to attach them to elements reachable from its Shadow DOM.

AFAIK, now it's something like

> Attaching CSS Variables' collection through Blending app, set them on `body` in light DOM. While setting it in view's composition, allows you to attach them to any element reachable from Shadow DOM of this compositions, so you can effectively scope their effect.

ErlendLandro commented on docs/topic-guides/blendable-web-apps/customizing-solutions-design.md:
```diff
@@ -0,0 +1,92 @@
+
+That is the lowest-level approach, that is slightly hacky, but this hackiness allows you to do any kind of changes and re-design.
+
+It allows you to completely change not only look but also the behavior of any UI element.
```
Why is this in its own paragraph?

tomalec replied:
as in https://github.com/Starcounter/Docs/pull/232#discussion_r192640252

ErlendLandro commented on docs/topic-guides/blendable-web-apps/customizing-solutions-design.md:
```diff
@@ -0,0 +1,92 @@
+
+{% hint style="info" %} The same technique works for any custom element, not only Uniform one. {% endhint %}
+
+## Need more?
```
Can we have a summary here that wraps up all the content and links to additional content? That's how we do it on other pages.

ErlendLandro commented on docs/topic-guides/blendable-web-apps/customizing-solutions-design.md:
```diff
@@ -0,0 +1,92 @@
+
+## Need more?
+
+If any of the above fulfill your needs for general customization and theming, first of all, please let us know.
```
Would it make sense to give people a way to let us know?

tomalec replied:
Definitely the question is what is this way, GH issue, SO question?

ErlendLandro replied:
Exactly, since it doesn't fit into a GitHub issue, I'm not sure. If we don't have a clear way for people to contact us about this, I suggest that we remove this sentence.

ErlendLandro commented on docs/topic-guides/blendable-web-apps/customizing-solutions-design.md:
```diff
@@ -0,0 +1,92 @@
+# Customizing Solution's Design
+
+
+Building your app or solution design the way we suggest does not only help you create a consistent look and feel fast and easily. It also gives you means to customize it later, by applying global themes, switching color & configuration schemes, or replacing the definition of every component at once.
```
Prefer "and" over "&"

Try to write the way that you speak, so "It also lets you" over "It also gives you means"

ErlendLandro reviewed (commented):
This is a hard topic to write about. I think you've made a good first effort. I made a lot of comments on style and wording.

If you want to improve the style even more, you can look at the [Microsoft](https://docs.microsoft.com/en-us/style-guide/top-10-tips-style-voice) and [Google](https://developers.google.com/style/tone) technical writing style guides.

You should also ask some of the app developers to look at it, to make sure that they understand.

tomalec committed:
Address style and wording comments from

@Mackiovello
https://github.com/Starcounter/Docs/pull/232#pullrequestreview-125447011

tomalec committed:
Apply more wording changes,

as suggested in https://github.com/Starcounter/Docs/pull/232#pullrequestreview-125447011

tomalec commented:
At https://github.com/Starcounter/uniform.css/issues/95#issuecomment-394892289 we decided to drop theming, so this PR should be either closed or refactored, please hold on, any reviews of it.

tomalec commented:
Given we dropped theming through Vaadin themes, what do you think about renaming our "CSS Custom Properties Collections" to "Themes"? At least for communication, so we could explain in the docs that it is technically a collection of CSS Custom Props, but UI will call it a theme for short.
@warpech , @Mackiovello

tomalec commented:
We agreed on our today's Web Platform Team meeting that we would like to reserve `theme` keyword for a feature that we would eventually implement. We all agreed that "CSS Custom Properties Collections" is a bad name, so we need to find a better alternative, but so far nothing new was proposed.

warpech committed:
rewrite big part of the design system page

- emphasize different layers of how Starcoutner host applies the common look and feel
- separate the roles of the app author and solution owner
- rename Starcoutner Design System to Uniform design system

warpech commented:
I am working on proposing some final changes to it.

warpech committed:
merge content of the "customizing solutions design" page into "uniform design system"

because after removing the outdated content about using Vaadin themes, that page became quite repetitive with "uniform design system"

warpech committed:
add a decision tree diagram

to illustrate different theming possibilities

warpech committed:
simplify the text

after I saw the wall of text in the rendering

warpech commented:
My changes are ready for re-review. @tomalec, @Mackiovello can you pls give a look?

tomalec commented on docs/topic-guides/blendable-web-apps/uniform-design-system.md:
```diff
@@ -2,17 +2,15 @@
-Starcounter host follows a layered design that allows controlling the look and feel of all apps. These layers are:
+Starcounter host follows a layered design that separates the responsibilities of the app author, the design system and the solution owner. Uniform fits in this design as the middle layer. 
+
+The full design consists of:
```
I don't get what you try to express here. Could you try to elaborate more, or call me on HO/slack?

tomalec commented on docs/topic-guides/blendable-web-apps/uniform-design-system.md:
```diff
@@ -0,0 +1,90 @@
+3. Theme configuration through CSS custom properties
+4. Extend or replace the theme with HTML, CSS, and JavaScript
+
+### Native HTML5 and Shadow DOM
```
Shadow DOM is part of HTML5, so what is this couple/layer, just HTML5?

tomalec commented on docs/topic-guides/blendable-web-apps/uniform-design-system.md:
```diff
@@ -0,0 +1,90 @@
+
+As an app author, you provide the views for your apps with a clear separation of the content and the presentation.
+
+In the content part, you provide the UI components that should be rendered by the Starcounter host as unstyled native HTML5 elements. 
```
Can't I use HTML4 elements? or some upcoming 5.4? I think the version does not matter here at all.

Also, as an application author, I'm not interested in any technology that renders my UI unstyled. 
"you provide the UI components that should be rendered by the Starcounter host as unstyled ... elements" gives me negative impression.

tomalec commented on docs/topic-guides/blendable-web-apps/uniform-design-system.md:
```diff
@@ -0,0 +1,90 @@
+
+In the content part, you provide the UI components that should be rendered by the Starcounter host as unstyled native HTML5 elements. 
+
+To prevent the Starcounter host from controlling the rendering of a UI component, you can use shadow DOM, preferably in a custom element. The shadow root creates a boundary that blocks the Starcounter host's attempt to apply the look and feel on your components automatically.
```
The beginning "To prevent the Starcounter host from controlling" seems disjointed, as you haven't mentioned before, that Starcounter takes over the control of the rendering.

tomalec replied:
Also, this paragraph seems to me like an exceptional case to do advanced things, and it's placed at the beginning before the basics were even briefly described.

tomalec commented on docs/topic-guides/blendable-web-apps/uniform-design-system.md:
```diff
@@ -0,0 +1,90 @@
+
+The CSS custom properties are the basic set of primitives. They define the fonts, sizes, colors, margins and other values used throughout the reset, patterns and components.
+
+The CSS reset, nicknamed [Underwear.css](https://starcounter.github.io/underwear.css/), overwrites the default user agent stylesheet. It provides the look of the native HTML5 elements such as the headings (H1-H6), links, inputs, buttons, and tables. It is implicitly loaded by the Starcounter app shell, so there is no need to load it explicitly in the views. 
```
I'd avoid 5 here as well. There is nothing 5-ish in `<h1>`

tomalec commented on docs/topic-guides/blendable-web-apps/uniform-design-system.md:
```diff
@@ -0,0 +1,90 @@
+
+The component library, [Uniform components](https://starcounter.github.io/uniform.css/components/), is a set of custom elements that implement interactive UI controls. For the beginning, we provide a form item, a date picker, a data table and a pagination control. 
+
+To use the patterns and the components, you must import them explicitly. They are placed in the shadow root of the view container to wrap the distributed content elements. That allows to serve real progressive enhancement, and what is crucial for a solution of blended apps: keep app content's markup semantic and unopinionated, giving solution owner a way to apply own opinion without changing the  ...
```
I'd say "They are meant to be used" instead of "They are placed" because they aren't, you as an author (app dev or solution owner) need to place it there.

tomalec commented on docs/topic-guides/blendable-web-apps/uniform-design-system.md:
```diff
@@ -0,0 +1,90 @@
+
+{% hint style="info" %} For more sophisticated theming we would love to use native [CSS shadow parts and themes](https://meowni.ca/posts/part-theme-explainer/). Unfortunately, it's not supported yet by any browser. {% endhint %}
+
+### Extend or replace the theme with HTML, CSS and JavaScript
```
Given we already agreed to reserve "theme" as a word for new features, I'd avoid it in this content. Would "presentation"  work, we already use it widely in our docs, so should not be confusing as introducing new keyword.

tomalec commented on docs/topic-guides/blendable-web-apps/uniform-design-system.md:
```diff
@@ -0,0 +1,90 @@
+
+As a solution owner, you are free to add, enhance, replace or hack the presentation layer using your HTML, CSS or JavaScript code. The only technical limitation is Shadow DOM encapsulation. 
+
+For example, when all apps consistently use Uniform components, all the date pickers are presented using `<uni-date-picker>`. You can drastically change its look or behavior by creating your own custom element.
```
I think that method is unrelated to shadow dom composition (previous paragraph) and you would not do that in Shadow DOM, you do that by replacing files on server side and works thanks to the fact that apps follow the convention of Uni-custom elements.

So this is not "extending and replacing theme/presentation with HTML, CSS and JavaScript", it's rather "extending and replacing HTML, CSS and JavaScript theme/presentation with your OS file manager and running new app". Those, in my opinion, should not be treated as one.

The feature that would be related to the previous paragraph is to replace the single instance of `uni-date-picker` with `other-party-date-picker`.

tomalec commented on docs/topic-guides/blendable-web-apps/uniform-design-system.md:
```diff
@@ -0,0 +1,90 @@
+
+To apply the new behavior just for some date pickers, you would fork `<uni-date-picker>` and give it a new name, such as `<your-date-picker>`. To use it, you would change `<uni-date-picker>` to `<your-date-picker>` in selected custom compositions.
+
+Create your own components by forking [`uni-element`](https://github.com/Starcounter/uniform.css/tree/master/components) or starting from scratch. In your components, you can use Uniform's CSS custom properties to apply common fonts, colors, etc.
```
This section describes in very detail very low level, heavy-weight solution, that's sometimes an overkill, comparing to simple tweaks or complete redesigns you can do with just shadow dom composition editing (which are not covered here so elaborately)

tomalec reviewed (commented):
Are you sure joining overview of UniDS and customizing it in one doc is a good idea? To me it makes it harder to diges and confusing, also it still misses some points.

I'd imagine it that way:
- Uniform-design-system.md
 - what it is,
 - why it's usefull,
 - how to use it, how to adhere to it,
 - how it's build
 - what it gives
   + (someting I found really missing in this version) a way to adjust the solution's presentation, easily and consistently with the rest of the system, without a need to overshoot with replacing custom elements definitions - recomposing, moving, adding few uni-elements and uni-patterns in the shadow dom

- Uniform theme configuration
  - variables (glabal, scoped, UI),
  - dreams about `::theme`,
  - HTML,CSS, JS freedom in compositions (code + visual editor)
  - replacing instances,
  - replacing CE definitions,
  - running forked SCF
  

BTW, I think the diagrams miss few paths: 
 - intersection about changing all vs. few
 - is it just about changing `Uniform components`? In my opinion, we should be talking about changing the look and feel of an entire solution of apps, not just uni.

tomalec commented:
I cannot "request changes" given for GH theoretically I'm the author ;)

warpech commented:
Me and Tomek have discussed some improvements today: https://docs.google.com/document/d/1n0FBnB6oOmWU1Bj9YP0e1tVTRBOSLYkJquQENr_WaSM/edit#heading=h.8rzw57a9qzhv. I will add them to the PR.

ErlendLandro commented:
@tomalec @warpech did you guys get anywhere with this?

warpech committed:
Merge branch '2.4' into uniform.css/issues/80-docs-about-theming

# Conflicts:
#	docs/topic-guides/blendable-web-apps/starcounter-design-system.md

warpech committed:
rename the page to "Common look and feel of apps"

because it is not only about Uniform but also about using something else than Uniform

warpech committed:
move the design kit information to UniformDocs

PR in UniformDocs: https://github.com/Starcounter/UniformDocs/pull/337

warpech committed:
make improvements based on the comments in https://docs.google.com/document/d/1n0FBnB6oOmWU1Bj9YP0e1tVTRBOSLYkJquQENr_WaSM/edit#

warpech commented:
Yes, would you mind reviewing? I have made all the changes that me and @tomalec discussed.

ErlendLandro commented on docs/topic-guides/blendable-web-apps/common-look-and-feel.md:
```diff
@@ -0,0 +1,84 @@
+
+### Common UI patterns and components of Uniform design system
+
+As the basic way for providing a common look and feel in the presentation layer, Starcounter provides Uniform design system. It consists of custom CSS properties, CSS reset and a library of patterns, and components.
```
>  It consists of custom CSS properties, CSS reset and a library of patterns, and components.

The comma placement is weird here. I usually use an oxford comma, with that, it should look like this:

>   It consists of custom CSS properties, CSS reset, and a library of patterns and components.

You could also skip the last comma:

>  It consists of custom CSS properties, CSS reset and a library of patterns and components.

Also, what does it mean to have a library of patterns? When I think patterns, I think OO patterns which are abstract descriptions of how to structure your code. What do we mean with patterns here? 

Is it a conscious choice to not use a definitive article for "Uniform design system"? Similar cases use definitive articles:
- [Carbon design system](http://www.carbondesignsystem.com/getting-started/designers): "The Carbon Design Kit is a living, breathing document that contains..."
- [Lighting design system](https://www.lightningdesignsystem.com/): "The Lightning Design System enables you to build..."

warpech replied:
> The comma placement is weird here. I usually use an oxford comma, with that, it should look like this

Thanks! I got tricked by Grammarly.

> When I think patterns, I think OO patterns which are abstract descriptions of how to structure your code. What do we mean with patterns here?

User Interface Design patterns are recurring solutions that solve common design problems. Design patterns are standard reference points for the experienced user interface designer. [User interface design patters](http://ui-patterns.com/patterns).

I will clarify in the text and reduce the usage of the word "pattern".

> Is it a conscious choice to not use a definitive article for "Uniform design system"? Similar cases use definitive articles:

Good catch! Yes, that *was* conscious, I wanted it to appear more flexible by that. but I agree this is clumsy. I will change to definitive.

ErlendLandro commented on docs/topic-guides/blendable-web-apps/common-look-and-feel.md:
```diff
@@ -0,0 +1,84 @@
+
+As the basic way for providing a common look and feel in the presentation layer, Starcounter provides Uniform design system. It consists of custom CSS properties, CSS reset and a library of patterns, and components.
+
+Uniform follows the inversion of control principle, which makes it a link between app authors and solution owners.
```
I don't understand this sentence. How is inversion of control related to links between app authors and solution owners?

warpech replied:
I will change it to: 

Uniform follows the inversion of control principle, which allows app authors to delegate the implementation and theming of common parts of user interface to the Starcoutner host and the solution owner.

ErlendLandro commented on docs/topic-guides/blendable-web-apps/common-look-and-feel.md:
```diff
@@ -0,0 +1,84 @@
+
+Unlike many design systems, Uniform is open-ended and themeable. It can be configured, extended or replaced in parts to achieve anything that is possible with HTML, CSS, and JavaScript.
+
+The custom CSS properties are the basic set of primitives. They define the fonts, sizes, colors, margins and other values used throughout the reset, patterns and components.
```
> The custom CSS properties are the basic set of primitives.

"basic set of primitives" is redundant. Primitives are always basic. 

I would phrase it something like this instead:

> The basic building blocks for the Uniform design system are custom CSS properties. 

Not sure if we need the word "custom" in there; aren't all CSS properties customizable?

ErlendLandro replied:
> Not sure if we need the word "custom" in there; aren't all CSS properties customizable?

Disregard this. they seem to be called custom CSS properties everywhere, so we should use it as well.

warpech replied:
"Basic building blocks" sound a little bit too cliche to me. Also, I would like Uniform components to be seen as building blocks. 

I will change it to "primitives". These things are sometimes called "design tokens", but we already use the word "token" for something else.

I will just change "the basic set of primitives" to "the set of primitives"

ErlendLandro commented on docs/topic-guides/blendable-web-apps/common-look-and-feel.md:
```diff
@@ -0,0 +1,84 @@
+
+The CSS reset overwrites the default user agent stylesheet. It provides the look of the native HTML5 elements such as the headings (H1-H6), links, inputs, buttons, and tables. It is implicitly loaded by the Starcounter app shell, so there is no need to load it explicitly in the views. 
+
+The pattern library, [Uniform.css](https://starcounter.github.io/uniform.css/), is a set of CSS classes that implement visual hierarchy and give specific roles to native elements. It provides sections, cards, titles, column layouts, alerts,  buttons, etc.
```
I don't understand this. What do you mean with "visual hierarchy" and how can you give native elements specific roles? Also, what's the last sentence trying to say? That the CSS classes work on that list of elements?

warpech replied:
Rephrased to:

The stylesheet part of Uniform implements visual hierarchy using sections, cards, and titles. It also allows giving specific roles to native elements, for example, to designate "primary" or "danger" buttons.

ErlendLandro commented on docs/topic-guides/blendable-web-apps/common-look-and-feel.md:
```diff
@@ -0,0 +1,84 @@
+
+The pattern library, [Uniform.css](https://starcounter.github.io/uniform.css/), is a set of CSS classes that implement visual hierarchy and give specific roles to native elements. It provides sections, cards, titles, column layouts, alerts,  buttons, etc.
+
+The component library, [Uniform components](https://starcounter.github.io/uniform.css/components/), is a set of custom elements that implement interactive UI controls. For the beginning, we provide a form item, a date picker, a data table and a pagination control. 
```
Can you try to rephrase  "For the beginning, we provide..."? 

Maybe something like: "The first version includes a form item, data picker, database table, and pagination control."

ErlendLandro commented on docs/topic-guides/blendable-web-apps/common-look-and-feel.md:
```diff
@@ -0,0 +1,84 @@
+
+The component library, [Uniform components](https://starcounter.github.io/uniform.css/components/), is a set of custom elements that implement interactive UI controls. For the beginning, we provide a form item, a date picker, a data table and a pagination control. 
+
+To use the patterns and the components, you must import them explicitly. They are placed in the shadow root of the view container to wrap the distributed content elements. That allows to serve real progressive enhancement, and what is crucial for a solution of blended apps: keep app content's markup semantic and unopinionated, giving solution owner a way to apply own opinion without changing the  ...
```
> They are placed in the shadow root of the view container to wrap the distributed content elements.

This sentence is Greek to me. I have no idea what it's supposed to mean. 

Overall, this whole paragraph is confusing to me. 
What's "real progressive enhancement"? What are you trying to say with "and what is crucial for a solution of blended apps:"?

Can you try to explain this section in other words to me and see if we can rephrase it?

warpech replied:
I agree this is too technical. I propose to rephrase it as:

To use the Uniform stylesheet and components, you must import them explicitly in the presentation layer of the view.

ErlendLandro commented on docs/topic-guides/blendable-web-apps/common-look-and-feel.md:
```diff
@@ -0,0 +1,84 @@
+
+The names of the custom CSS properties, CSS classes, and custom elements begin with a `uni-` prefix.
+
+Our [UniformDocs](https://uniform.starcounter.io/) app demonstrates the use of Uniform design system as well as commonly used 3rd party custom elements that we can endorse, such as a Google map or a combo box.
```
I think you should divide this paragraph into two sentences:

> Our [UniformDocs](https://uniform.starcounter.io/) app demonstrates the use of Uniform design system. It also shows how to use common third-party custom elements such as the Google Map or Vaadin combo box elements.

ErlendLandro commented on docs/topic-guides/blendable-web-apps/common-look-and-feel.md:
```diff
@@ -0,0 +1,84 @@
+
+Our [UniformDocs](https://uniform.starcounter.io/) app demonstrates the use of Uniform design system as well as commonly used 3rd party custom elements that we can endorse, such as a Google map or a combo box.
+
+As an app author, you are invited to implement specialized components in your app that use Uniform internally. This will allow to adapt to the solution’s look and feel also for your specialized components.
```
"you are invited to" => "you can"

> This will allow to adapt to the solution’s look and feel also for your specialized components.
I would change this slightly:

> This lets you adapt specialized components to the solution’s look and feel.

ErlendLandro commented on docs/topic-guides/blendable-web-apps/common-look-and-feel.md:
```diff
@@ -0,0 +1,84 @@
+
+Uniform design system comes with a modern-looking default theme.
+
+As a solution owner, you can customize it by providing new values for the [custom CSS properties](https://developer.mozilla.org/en-US/docs/Web/CSS/Using_CSS_variables). This allows changing the fonts, colors, etc. without replacing the stylesheet.
```
We're talking about custom CSS properties earlier, but this is the first time we link to it. Only the first mention should be a link. 

In general, avoid using "etc." Both the [Microsoft](https://docs.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/e/etc) and [Google](https://developers.google.com/style/word-list#letter-E) technical writing style guides advice against it.

ErlendLandro commented on docs/topic-guides/blendable-web-apps/common-look-and-feel.md:
```diff
@@ -0,0 +1,84 @@
+
+You can provide new values directly in a stylesheet in a custom composition. It requires some skill in CSS and HTML but allows for spectacular results such as providing a theme configuration only for a specific part of the app.
+
+If you don't need that much of control and prefer a simple to use GUI, use the "Theme configurations" feature of [Blending](https://github.com/Starcounter/Blending) app suite. It is an administration tool that stores the values of custom CSS properties in the database. It can provide theme configurations for specific layouts that affect all Uniform components on the screen.
```
"the Blending app suite" 

"affect all the Uniform components"

ErlendLandro reviewed (commented):
This looks comprehensive. Anyway, for someone like me that's not doing web development daily, a lot of things here are confusing. 

I also suggested some wording changes. 

I'm missing some kind of visualization. I have a hard time understanding the hierarchy of the different styling options right now. A visualization could really help with that.

ErlendLandro commented:
One more thing: it would be nice to have a single paragraph on the bottom of the page to summarize.

ErlendLandro commented:
Since I'm not touching the web stuff anymore, I think it would be good to get the perspective of someone in the apps team since they are closer to this. Maybe @Mihaiii or @Joltrix

warpech committed:
fixes according to @Mackiovello's comments in https://github.com/Starcounter/Docs/pull/232

warpech commented:
Thank you for a very detailed review. I have reacted to all your comments.

@Mihaiii's or @Joltrix's comments are very welcome, but this page is not the most important in the world. It is important to finally ship this PR one way or another. If you agree, I will merge for now.

ErlendLandro commented on docs/topic-guides/blendable-web-apps/common-look-and-feel.md:
```diff
@@ -23,43 +23,43 @@ As an app author, you provide the views for your apps with a clear separation of
+As the basic way for providing a common look and feel in the presentation layer, Starcounter provides [Uniform Design System](https://uniform.starcounter.io/). It consists of theme primitives, CSS reset, and a library of user interface design patterns implemented using a stylesheet and components.
 
-Uniform follows the inversion of control principle, which makes it a link between app authors and solution owners.
+Uniform follows the inversion of control principle, which allows app authors to delegate the implementation and themeing of common parts of user interface to the Starcoutner host and the solution owner.
```
Typo "Starcoutner"  and "themeing" -> "theming"

ErlendLandro commented on docs/topic-guides/blendable-web-apps/common-look-and-feel.md:
```diff
@@ -23,43 +23,43 @@ As an app author, you provide the views for your apps with a clear separation of
 {% hint style="warning" %} Attaching custom CSS properties through "Theme configurations" sets them on the global scope - the `body` of the light DOM (https://github.com/Starcounter/Blending/issues/232). Providing custom CSS properties values in custom compositions attaches them in shadow DOM to the scope that you choose. {% endhint %}
 
-{% hint style="info" %} For more sophisticated theming we would love to use native [CSS shadow parts and themes](https://meowni.ca/posts/part-theme-explainer/). Unfortunately, it's not supported yet by any browser. {% endhint %}
+{% hint style="info" %} For more sophisticated themeing we would love to use native [CSS shadow parts and themes](https://meowni.ca/posts/part-theme-explainer/). Unfortunately, it's not supported yet by any browser. {% endhint %}
```
Why did you change spelling to "themeing"? Others spell it "theming", see, for example [material design](https://material.io/design/material-theming/overview.html#material-theming)

warpech replied:
Late hour... Thanks for noticing, I will fix

ErlendLandro commented on docs/topic-guides/blendable-web-apps/common-look-and-feel.md:
```diff
@@ -73,12 +73,12 @@ To apply the new behavior for all date pickers in the solution, you would simply
+The following diagram presents a decision tree of what technique of themeing to use in your case.
 
-![Decision tree of theming Uniform design system](../../.gitbook/assets/uniform-decision-tree.png)
+![Decision tree of themeing Uniform](../../.gitbook/assets/uniform-decision-tree.png)
```
Same here

ErlendLandro reviewed (approved):
Fix these typos and you can merge

warpech committed:
fix typos

should be: Starcounter, theming

warpech commented:
Thanks once again! Merging
